teps/0145-cel-in-whenexpression.md

```diff
@@ -0,0 +1,336 @@
+
+Take a look at current [WhenExpressions]:
+
+```yaml
+when:
+  - input: "$(params.path)"
+    operator: in
+    values: ["README.md"]
+```
+
+Current `WhenExpression` requires users to define three fields: `input`, `operator` and `values`. but CEL expressions can achieve the same functionality with a single string. Additionally, the current operator only supports `in` and `notin`, which can be limiting for complex use cases. For example, users will need to figure out [hacky workarounds][#3591] to support any in their case.
+
+For these reasons, `WhenExpressions` would be a good place to start introducing CEL in Tekton Pipelines. This would allow users to write more concise and powerful conditional expressions, and it could be extended to other fields in the future.
+
+### Use Cases
+
+- Cover all the functionalities of current `WhenExpression` and offer a simpler way of having the conditional execution
+- Support ANY: e.g. Execute the task with if any params is not empty
+- Support numeric operators such as GreaterThan/LessThan, e.g. Execute the task if test coverage is large than a certain number.
+
+### Goals
+- Support CEL in `WhenExpression`.
+
+### Non-Goals
+- Support CEL in Pipeline API fields other than `WhenExpression`.
+
+### Related Work
+
+Similar expression evalutation has been supported in other CI/CD tools.
+- [Argo][Argo]: `Workflows` support conditional execution using a when property to specify whether to run or skip a `Step` ([example][argo example]).
+- [GitHub Actions][GitHub Actions]: supports evaluating expressions in `if` of a `step`.
+- [Jenkins][Jenkins]: executes the stage when the specified Groovy expression evaluates to true.
+- [Spinnaker][Spinnaker]: Uses string expressions, `Stages` only run when expressions evaluate to `True`.
+
+
+## Proposal
+
+We propose to add a `cel` field under `WhenExpression` and the type is `string`. Same as [CelCustomRun][CELCustomRun], we propose to use [cel-go][cel-go] for the expression evaluation. So users can craft any valid CEL expression as defined by the [cel-spec][cel-spec] language definition.
+
+
+### Syntax
+
+In this section we will show how CEL can support the [use cases](#use-cases).
+
+#### Use CEL to replace current WhenExpression
+```yaml
+# current WhenWxpressions
+when:
+  - input: "foo"
+    operator: "in"
+    values: ["foo", “bar”]
+  - input: "duh"
+    operator: "notin"
+    values: ["foo", “bar”]
+
+# with cel
+when:
+  - cel: "\”foo\” in [\"foo\", \“bar\”]”
+  - cel: "\”duh\” not in [\"foo\", \“bar\”]”
+```
+
+**Note:** The `WhenExpressions` contains a list of `WhenExpression`, if any `WhenExpression` is false it will return false. So we’re using ANY to evaluate the list of `WhenExpressions`. And this will remain the same for this proposal. For each WhenExpression, there will be one CEL expression which can fully replace the current when expression functionalities.
```

Comment:
```suggestion
**Note:** The `WhenExpressions` contains a list of `WhenExpression`, if all `WhenExpressions` are true it will return true and allow execution. So we’re using AND to evaluate the list of `WhenExpressions`. And this will remain the same for this proposal. Each existing `WhenExpression` can be replaced with a CEL expression.
```